File: leetcode/27-remove-element/27-remove-element.js
/**
 * @param {number[]} nums
 * @param {number} val
 * @return {number}
 */
var removeElement = function(nums, val) {
  for (let i = nums.length - 1; i >= 0 ; i--) {
    if (nums[i] == val) {
       nums.splice(i, 1);
    }
  }
  return nums.length;
};


let test = function(q, t, a) {
  const sol = removeElement(q, t);
  console.assert(sol === a.length, `Fail: length, ${sol} !== ${a.length}`);
  for (let i = 0; i < sol; i++) {
    console.assert(q[i] === a[i], `Fail: index ${i}: ${q[i]} !== ${a[i]}`);
  }
};

const q1 = [3, 2, 2, 3];
const t1 = 3;
const a1 = [2, 2];
test(q1, t1, a1);

const q2 = [0, 1, 2, 2, 3, 0, 4, 2];
const t2 = 2;
const a2 = [0, 1, 3, 0, 4]
test(q2, t2, a2);

File: leetcode/27-remove-element/27-remove-element.py
class Solution:
    def removeElement(self, nums: 'list[int]', val: 'int') -> 'int':
        l = []
        for i in range(len(nums)):
            if nums[i] != val:
                l.append(nums[i])

        for i in range(len(l)):
            nums[i] = l[i]
        return len(l)


# class Solution:
#     def removeElement(self, nums: 'list[int]', val: 'int') -> 'int':
#         for i in range(len(nums)-1, -1, -1):
#             if nums[i] == val:
#                 nums.pop(i)
#         return len(nums)


if __name__ == '__main__':
    sol = Solution()
    l1 = [3, 2, 2, 3]
    assert sol.removeElement(l1, 3) == 2, 'Fail'
    assert l1[0] == 2, 'Fail'
    assert l1[1] == 2, 'Fail'

    l2 = [0, 1, 2, 2, 3, 0, 4, 2]
    assert sol.removeElement(l2, 2) == 5, 'Fail'
    assert l2[0] == 0, 'Fail'
    assert l2[1] == 1, 'Fail'
    assert l2[2] == 3, 'Fail'
    assert l2[3] == 0, 'Fail'
    assert l2[4] == 4, 'Fail'
